### ISSUE TITLE
Retry signing of blocks after re-starting the client

### ISSUE THREAD

**Issue by devos50** [Tribler/py-ipv8#724]
Our [recent SQLite error](https://github.com/Tribler/py-ipv8/pull/720) caused the creation of many unsigned blocks, which will most likely never be signed with the current implementation of TrustChain.

I propose to implement a better block signing retry mechanism, where a peer more actively keeps track of the blocks that are yet to be countersigned by a party, and attempts to get these blocks signed, even after restarting the software.

**Comment by qstokkink** [comment]
Won't the counterparty have lost the values in memory (say a week later) that would make it O.K. to countersign the block?

Also, this has the potential to decimate disk I/O and performance.

**Comment by devos50** [comment]
> Won't the counterparty have lost the values in memory (say a week later) that would make it O.K. to countersign the block?

That depends on whether signing of the block needs some state that's only available shortly after creation of the block. I would say this depends on the application layer.

> Also, this has the potential to decimate disk I/O and performance.

The question here is how reliably we want TrustChain to be. Counterparties being offline (or refusing to counter-sign a block due to a bug, as we have seen) is a realistic scenario and a retry mechanism could help here to avoid blocks from never being signed. And yes, this most likely impacts performance.

**Comment by qstokkink** [comment]
> I would say this depends on the application layer.

So, why not leave this to the application layer then?

> Counterparties being offline (or refusing to counter-sign a block due to a bug, as we have seen) is a realistic scenario and a retry mechanism could help here to avoid blocks from never being signed.

Well not unless the bug is fixed: you would need long-term application-specific persistent memory to solve this and have it kick in after a protocol update. The only realistic thing I can see nodes recovering from with a retry is temporary network outage, without restarting the process. This is already covered by the current implementation.

In cases where Trustchain serves as a commitment to a future action (market + identity) which times out rather quickly anyway (for example identity attributes have a window of 5 minutes), this would just serve to slow down everything.

Long story short: I am not opposed to providing a hook to override retry functionality from within IPv8. BUT, it should be application specific (as it also requires additional logic) and not enabled by default (the naive retry we have now is sufficient for the default case).

**Comment by ichorid** [comment]
I agree with @qstokkink . According to [end-to-end principle](https://devopedia.org/end-to-end-principle), the decision about how to handle a failure should be done by the application layer, not the middleware.

However, I would argue that adding an optional callback will be harmful because it will be not obvious for a programmer that it actually *exists*. Instead, we should just `raise` an error and let the programmer catch it.

**Comment by qstokkink** [comment]
> Instead, we should just raise an error and let the programmer catch it.

When the `HalfBlockSignCache` times out? We could use the new managed future feature to set the exception on a Future -- but someone would actually have to await the future. A callback seems more practical (and less error-prone).

**Comment by ichorid** [comment]
I don't mind the callback. But then the default implementation of the callback should raise `NotImplemented` error, so the programmer will be notified that something went wrong.

**Comment by qstokkink** [comment]
It seems like this issue will not be resolved as suggested in the O.P.

As we're also migrating Trustchain out of IPv8, I'll close this issue.
